app/models/sale.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+class Sale < ApplicationRecord
+  belongs_to :seller, class_name: 'User', optional: true
+  belongs_to :client, class_name: 'User'
+  belongs_to :payment_method
+  belongs_to :invoice
+  has_one :subscription, dependent: :destroy
+  has_many :refunds, dependent: :destroy
+  has_many :articles_sales, dependent: :destroy
+  has_many :articles, through: :articles_sales
+  has_many :sales_subscription_offers, dependent: :destroy
+  has_many :subscription_offers, through: :sales_subscription_offers
+
+  accepts_nested_attributes_for :articles_sales
+  accepts_nested_attributes_for :sales_subscription_offers
+
+  validates :total_price, presence: true
+
+  # before_create :create_associated_subscription
+
+  def verify
+    self.verified_at = Time.zone.now if verified_at.nil?
+  end
+
+  def generate(duration:, seller:)
+    generate_sales_subscription_offers duration.to_i
+    self.seller = seller
+    update_total_price
+    verify if payment_method.auto_verify
+    generate_invoice
+  end
+
+  def update_total_price
+    total = 0
+    articles_sales.each do |rec|
+      total += rec.quantity * Article.find(rec.article_id).price
+    end
+    sales_subscription_offers.each do |rec|
+      total += rec.quantity * SubscriptionOffer.find(rec.subscription_offer.id).price
+    end
+    self.total_price = total
+  end
```

Comment:
it's a bit of a code smell here because we need to call `update_total_price` before getting the total price. we could compute this value every time we get the total price instead, we would have the guarantee always to have the correct total price.
one benefit of being explicit on the update of the total price is that we don't compute it every time we get the total price but we don't need this kind of performance optimization (and we would need to adapt the logic anyway to guarantee that the total price is always up-to-date)

Reply:
We call for `update_total_price` when there is no subscription offers nor articles to be add to the sale so there will not have any update of the price after. I agree that is not quite convincing but I did not find anything else to do so.

Reply:
Do we need to have the total price in the database? The total price can be calculated at runtime by summing the price of articles and subscriptions, keeping it as a column in the database is redundant and we might have consistency issues (the total price of the sale is different from the sum of the articles' and subscriptions' price because we forgot to update the total price somewhere). And we can change `update_total_price` to be a getter (renaming it to `total_price` and returning the computed total price)

Reply:
The fact is each time we would like to show a sale we need to calculate again the total price of the sale. That would be redundant because once a sale is created we want it to be immutable so after the creation the price couldn't be changed so there is no need to update the price later we just have to calculate it when we create it and it enables to avoid recalculate it each time we want to display which is the case when we show a user, we'd like to show every sales and their total price.

Reply:
I don't think recalculating it every time is an issue. We certainly don't need that kind of optimization (we wouldn't even notice it). Making the code more resilient to bugs is more important in my opinion.

Regarding storage issues, we have a redundancy with the pdf if we need to know the price. I would say it's good enough to make sure we can't lose the invoices.

My comment is not blocking in any way, I don't mind keeping it like this